neuralmonkey/attention/scaled_dot_product.py

```diff
@@ -0,0 +1,176 @@
+                             .format(self._dimension, self.n_heads))
+
+        self._head_dim = int(self._dimension / self.n_heads)
+        self._scaling_factor = 1 / math.sqrt(self._head_dim)
+
+    def attention(self,
+                  query: tf.Tensor,
+                  decoder_prev_state: tf.Tensor,
+                  decoder_input: tf.Tensor,
+                  loop_state: MultiHeadLoopStateTA,
+                  step: tf.Tensor) -> Tuple[tf.Tensor, MultiHeadLoopStateTA]:
+
+        if self.n_heads == 1:
+            context, weights = self.attention_single_head(
+                query, self.attention_keys, self.attention_values)
+            head_weights = [weights]
+        else:
+            # project query, keys and vals: [batch, rnn] to [batch, rnn2]
+            head_contexts, head_weights = zip(*[
+                self.attention_single_head(
+                    tf.layers.dense(query, self._head_dim,
+                                    name="query_proj_head{}".format(i)),
+                    tf.layers.dense(self.attention_keys, self._head_dim,
+                                    name="keys_proj_head{}".format(i)),
+                    tf.layers.dense(self.attention_values, self._head_dim,
+                                    name="values_proj_head{}".format(i)))
+                for i in range(self.n_heads)])
+
+            context = tf.layers.dense(
+                tf.concat(head_contexts, -1), self._dimension,
+                name="output_proj")
+
+        next_contexts = loop_state.contexts.write(step, context)
+        next_head_weights = [loop_state.head_weights[i].write(step,
+                                                              head_weights[i])
+                             for i in range(self.n_heads)]
+
+        next_loop_state = MultiHeadLoopStateTA(
+            contexts=next_contexts,
+            head_weights=next_head_weights)
+
+        return context, next_loop_state
+
+    def attention_single_head(self, query: tf.Tensor,
+                              keys: tf.Tensor,
+                              values: tf.Tensor) -> Tuple[tf.Tensor,
+                                                          tf.Tensor]:
+        # shape: batch, time (similarities of attention keys in batch and time
+        # to the queries in the batch)
+        dot_product = tf.reduce_sum(
+            tf.expand_dims(query, 1) * keys, [-1])
+        energies = dot_product * self._scaling_factor
+
+        weights = tf.nn.softmax(energies)
+
+        if self.attention_mask is not None:
+            weights_all = weights * self.attention_mask
+            norm = tf.reduce_sum(weights_all, 1, keep_dims=True) + 1e-8
+            weights = weights_all / norm
+
+        # apply dropout to the weights (Attention Dropout)
+        weights = dropout(weights, self.dropout_keep_prob, self.train_mode)
+
+        # sum up along the time axis
+        context = tf.reduce_sum(
+            tf.expand_dims(weights, -1) * values, [1])
```

Comment:
Nebylo by lepsi tohle nahradit (podobne i na jinych mistech, kde delas dot product)?
context = tf.matmul(tf.expand_dims(weights, -1), values)

Nebo je tvoje reseni efektivnejsi?

Reply:
Tohle hlavně není dot product, ale weighted average. Poslední dimenze vah se broadcastuje a pak se to sečte.

Jsou tam místa, kde se podobně píše dot product, většina z nich jako dotproduct napsat nejde, protože to o dimenzi nesedí.

Reply:
Já myslim, že @varisd má pravdu. Pokud děláš matmul matic s tvarem [1, N] a [N, M], dostaneš (vodorovnej) vektor [1, M], kde každej prvek na tom řádku bude skalární součin toho sloupcovýho vektoru s korespondujícím řádkem tý druhý matice. Což odpovídá váženýmu součtu čísel z odpovídajícího "kanálu" přes všechny časy.

Jediný co je na @varisd  návrhu špatně je ta mínus jednička, místo který patří podle mě nula.

Reply:
Jo, urcite tam ma byt nula.